src/mapstorage/XmlMapStorage.cpp

```diff
@@ -41,20 +214,396 @@ void XmlMapStorage::newData()
+        } else if (name == "description") {
+            room.setDescription(RoomDesc{loadString(stream)});
+        } else if (name == "exit") {
+            loadExit(stream, exitList);
+        } else if (name == "light") {
+            room.setLightType(loadEnum<RoomLightEnum>(stream));
+        } else if (name == "loadflag") {
+            loadFlags |= loadEnum<RoomLoadFlagEnum>(stream);
+        } else if (name == "mobflag") {
+            mobFlags |= loadEnum<RoomMobFlagEnum>(stream);
+        } else if (name == "note") {
+            room.setNote(RoomNote{loadString(stream)});
+        } else if (name == "portable") {
+            room.setPortableType(loadEnum<RoomPortableEnum>(stream));
+        } else if (name == "ridable") {
+            room.setRidableType(loadEnum<RoomRidableEnum>(stream));
+        } else if (name == "sundeath") {
+            room.setSundeathType(loadEnum<RoomSundeathEnum>(stream));
+        } else if (name == "terrain") {
+            room.setTerrainType(loadEnum<RoomTerrainEnum>(stream));
+        } else {
+            qWarning().noquote().nospace()
+                << "At line " << stream.lineNumber() << ": ignoring unexpected XML element <"
+                << name << "> inside <room id=\"" << idstr << "\">";
+        }
+        skipXmlElement(stream);
+    }
+    room.setExitsList(exitList);
+    room.setLoadFlags(loadFlags);
+    room.setMobFlags(mobFlags);
+
+    loadedRooms.emplace(std::move(roomId), std::move(sharedroom));
+}
+
+// convert string to RoomId
+RoomId XmlMapStorage::loadRoomId(QXmlStreamReader &stream, const QStringRef &idstr)
+{
+    const RoomId id{idstr.toUInt()};
+    // convert number back to string, and compare the two:
+    // if they differ, room ID is invalid.
+    if (idstr != roomIdToString(id)) {
+        throwErrorFmt(stream, "invalid room id \"%1\"", idstr);
+    }
+    return id;
+}
+
+// load current <coord> element
+Coordinate XmlMapStorage::loadCoordinate(QXmlStreamReader &stream)
+{
+    const QXmlStreamAttributes attrs = stream.attributes();
+    bool fail = false;
+    const int x = conv.toNumber<int>(attrs.value("x"), fail);
+    const int y = conv.toNumber<int>(attrs.value("y"), fail);
+    const int z = conv.toNumber<int>(attrs.value("z"), fail);
+    if (fail) {
+        throwErrorFmt(stream,
+                      "invalid coordinate values x=\"%1\" y=\"%2\" z=\"%3\"",
+                      attrs.value("x").toString(),
+                      attrs.value("y").toString(),
+                      attrs.value("z").toString());
+    }
+    return Coordinate(x, y, z);
+}
+
+// load current <exit> element
+void XmlMapStorage::loadExit(QXmlStreamReader &stream, ExitsList &exitList)
+{
+    const QXmlStreamAttributes attrs = stream.attributes();
+    const ExitDirEnum dir = directionForLowercase(attrs.value("dir"));
+    DoorFlags doorFlags;
+    ExitFlags exitFlags;
+    Exit &exit = exitList[dir];
+    exit.setDoorName(DoorName{attrs.value("doorname").toString()});
```

Comment:
Luckily XML as a file format is recent enough that it automatically handles encoding woes (quite) sanely.

The first line of XML files generated with "Export MMapper2 Base Map As..."  always contains the following:
```
<?xml version="1.0" encoding="UTF-8"?>
```
Users can manually change that encoding in their own saved XML files if they want,
and QXmlStreamReader should honour whatever encoding is written in the first line when loading an XML map,
automatically converting it internally to QString and/or QStringRef (which are UTF-16).

TaggedString<> and its aliases (DoorName, InfoMarkText, RoomName, RoomDesc, RoomContents, RoomNote)
automatically do the necessary (and as correct as possible) conversion from UTF-16 to Latin1 in the constructor
```
explicit TaggedString(const QString &s)
        : m_str(::toStdStringLatin1(s))
    {}
```
In summary, the code is already correct.

I tested loading an XML map containing the non-ascii characters ÀÈÌÒÙ ÁÉÍÓÚ àèìòù áéíúó
(encoded as UTF-8 in the file, since it started with encoding="UTF-8"), and they were shown correctly in the GUI.

If you write in the file some characters **not** representable in Latin1, as for example the euro sign €,
they will be converted to '?' by the `::toStdStringLatin1(s)` called by `TaggedString` constructor.

This is the best we can currently do, because MMapper2 internally uses Latin1 for most strings.
If it used UTF-8 or UTF-16 (never mind UTF-32), loading/saving any unicode character from/to XML would be perfect.

Reply:
LGTM, thanks for explaining